### PR TITLE
Feature metrics shm

### PR DESCRIPTION
**Pull request by neverhook**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Sign the CLA
Make sure you have signed the [CLA](https://www.clahub.com/agreements/alipay/sofa-mosn)

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, 
details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases.
And you need to show the ut result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/metrics/shm/zone.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+		}
+	}
+}
+
+func (z *zone) unlock() {
+	times := 0
+
+	// 5ms spin interval, 5 times burst
+	for true {
+		if atomic.CompareAndSwapUint32(z.mutex, pid, 0) {
```
Comment (taoyuanyuan) [inline pkg/metrics/shm/zone.go]: 感觉不会存在这种情况,  如果不成功, 就是使用有问题吧

File: pkg/metrics/shm/entry.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+func (e *metricsEntry) assignName(name []byte) {
+	i := 0
+	for i = range name {
+		e.name[i] = name[i]
+	}
+	e.name[i+1] = 0
+}
+
+func (e *metricsEntry) equalName(name []byte) bool {
```
Comment (taoyuanyuan) [inline pkg/metrics/shm/entry.go]: 可以用 bytes.Equal

File: pkg/metrics/gc_pause_test.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+	ResetAll()
+	gcPause()
+
+	//------------------------------------------
+
+	func() {
+		zone := shm.InitMetricsZone("testGCPause", 20000000*256)
+		defer zone.Detach()
+
```
Comment (taoyuanyuan) [inline pkg/metrics/gc_pause_test.go]: 需要增加ResetAll(), 不然会导致后面的测试panic

File: pkg/metrics/shm/hashset.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+	offset += 4 * slotsNum // slot type is uint32
+	if offset > bytesNum {
+		return nil, errors.New("segment is not enough to map hashSet.slots")
+	}
+
+	// 4. initialize
+	// 4.1 slots
+	for i := 0; i < slotsNum; i++ {
+		set.slots[i] = sentinel
```
Comment (taoyuanyuan) [inline pkg/metrics/shm/hashset.go]: 如果是升级新启动的mosn, 这儿不是会覆盖共享内存中的值吗?

File: pkg/metrics/shm/hashset.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+	}
+
+	// 4.2 entries
+	last := cap - 1
+	for i := 0; i < last; i++ {
+		set.entry[i].next = uint32(i + 1)
+	}
+	set.entry[last].next = sentinel
```
Comment (taoyuanyuan) [inline pkg/metrics/shm/hashset.go]: 同上

File: pkg/metrics/shm/zone.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+	atomic.AddUint32(zone.ref, 1)
+
+	return zone, nil
+}
+
+func (z *zone) lock() {
+	times := 0
+
+	// 5ms spin interval, 5 times burst
+	for true {
```
Comment (taoyuanyuan) [inline pkg/metrics/shm/zone.go]: true 去掉

File: pkg/metrics/shm/zone.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+		mutex: (*uint32)(unsafe.Pointer(mutex)),
+		ref:   (*uint32)(unsafe.Pointer(ref)),
+	}
+	// add ref
+	atomic.AddUint32(zone.ref, 1)
+
+	return zone, nil
+}
+
+func (z *zone) lock() {
```
Comment (taoyuanyuan) [inline pkg/metrics/shm/zone.go]: 需要增加防止死锁的逻辑

File: pkg/types/config.go [hunk 8]
```diff
@@ -37,6 +37,7 @@ var (
 	TransferConnDomainSocket   = MosnConfigPath + string(os.PathSeparator) + "conn.sock"
 	TransferStatsDomainSocket  = MosnConfigPath + string(os.PathSeparator) + "stats.sock"
 	TransferListenDomainSocket = MosnConfigPath + string(os.PathSeparator) + "listen.sock"
+	TransferMetricsDomainSocket = MosnConfigPath + string(os.PathSeparator) + "metrics.sock"
```
Comment (taoyuanyuan) [inline pkg/types/config.go]: 无用代码

File: pkg/types/config.go [hunk 9]
```diff
@@ -69,6 +70,7 @@ func InitDefaultPath(path string) {
 	TransferConnDomainSocket = MosnConfigPath + string(os.PathSeparator) + "conn.sock"
 	TransferStatsDomainSocket = MosnConfigPath + string(os.PathSeparator) + "stats.sock"
 	TransferListenDomainSocket = MosnConfigPath + string(os.PathSeparator) + "listen.sock"
+	TransferMetricsDomainSocket = MosnConfigPath + string(os.PathSeparator) + "metrics.sock"
```
Comment (taoyuanyuan) [inline pkg/types/config.go]: 同上

File: pkg/metrics/shm/entry.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+// metricsEntry is the mapping for metrics entry record memory-layout in shared memory.
+//
+// This struct should never be instantiated.
+type metricsEntry struct {
+	value int64     // 8
+	ref   uint32    // 4
+	name  [100]byte // 100
+}
+
+func (e *metricsEntry) assignName(name []byte) {
```
Comment (taoyuanyuan) [inline pkg/metrics/shm/entry.go]: len 判断
